src/Forms/Container.php

```diff
@@ -56,53 +66,88 @@ public function setDefaults(iterable $values, bool $erase = false)
 	 * @return static
 	 * @internal
 	 */
-	public function setValues(iterable $values, bool $erase = false)
+	public function setValues($data, bool $erase = false)
 	{
-		if ($values instanceof \Traversable) {
-			$values = iterator_to_array($values);
-
-		} elseif (!is_array($values)) {
-			throw new Nette\InvalidArgumentException(sprintf('First parameter must be an array, %s given.', gettype($values)));
+		$components = [];
+		foreach ($this->getComponents() as $name => $control) {
+			if ($control instanceof IControl || $control instanceof self) {
+				$components[$name] = $control;
+			}
 		}
 
-		foreach ($this->getComponents() as $name => $control) {
-			if ($control instanceof IControl) {
-				if (array_key_exists($name, $values)) {
-					$control->setValue($values[$name]);
+		if (is_object($data)) {
+			$this->mappedClass = get_class($data);
+			$mapper = $this->mapper ?: new Mapper;
+			$data = $mapper->objectToValues($data, array_keys($components));
 
-				} elseif ($erase) {
-					$control->setValue(null);
-				}
+		} elseif (!is_array($data)) {
+			throw new Nette\InvalidArgumentException(sprintf('First parameter must be an array or object, %s given.', gettype($data)));
+		}
 
-			} elseif ($control instanceof self) {
-				if (array_key_exists($name, $values)) {
-					$control->setValues($values[$name], $erase);
+		foreach ($components as $name => $control) {
+			if (array_key_exists($name, $data)) {
+				if ($control instanceof IControl) {
+					$control->setValue($data[$name]);
+				} else {
+					$control->setValues($data[$name] ?: [], $erase);
+				}
 
-				} elseif ($erase) {
+			} elseif ($erase) {
+				if ($control instanceof IControl) {
```

Comment:
I don't like those conditions. How about we split `$components` array into `$controls`, `$containers` and `$dataKeys`.